### PR TITLE
Common widget GzColor (Citadel)

### PR DESCRIPTION
**Pull request by AzulRadio**
#  New feature

* Part of #310

## Summary

Pack RGB color GUI into a common widget `GzColorRGB.qml`

Replace the color GUI in `Grid3d` with the common widget

## Test it

(New)
demo_grid3d_new

(Deprecated)
demo_gzcolor_grid3d

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: include/ignition/gui/qml/GzColorRGB.qml [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      radius: 5
+      border.color: Qt.rgba(r / 255, g / 255, b / 255, a)
+      border.width: 2
+      color: Qt.rgba(r / 255, g / 255, b / 255, a)
+    }
+    onClicked: colorDialog.open()
+
+    ColorDialog {
+      id: colorDialog
+      title: "Choose a grid color"
```
Comment (chapulina) [inline include/ignition/gui/qml/GzColorRGB.qml:141]:
```suggestion
      title: "Choose a color"
```

This widget will be used for other things besides the grid

**Review by chapulina — CHANGES_REQUESTED** [review]
Looking good! Just a few comments

File: include/ignition/gui/qml/GzColorRGB.qml [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    Layout.row: 0
+    Layout.column: 5
+    text: "Alpha"
+    color: "dimgrey"
+    visible: textVisible
+  }
+
+  IgnSpinBox {
+    id: "a_spin"
+    maximumValue: 1.0
```
[the remark this answers is not in the document]
Comment (AzulRadio) [reply include/ignition/gui/qml/GzColorRGB.qml]: Set alpha to 0 - 1.0 for 2 reasons:
1. Many RGBA standards use 0 - 255 int for RGB and 0 - 1.0 float for A. For example: CSS (https://www.w3schools.com/cssref/func_rgba.asp) and HTML (https://www.w3schools.com/html/html_colors_rgb.asp) RGBA. So many online RGBA color pickers also implement the same idea (https://rgbacolorpicker.com/). Keeping alpha 0-1.0 can save users from another conversion.
2. Alpha is the transparency and is usually used with the following color blend formula `alpha * Color1 + (1 - alpha) * Color2`. Setting it to 0 - 1.0 can be more intuitive and comfortable for users.

File: include/ignition/gui/qml/GzColor.qml [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+ *
+*/
+import QtQuick 2.9
+import QtQuick.Controls 2.1
+import QtQuick.Dialogs 1.0
+import QtQuick.Layouts 1.3
+
+
+GridLayout {
+  id: root
```
Comment (jennuine) [inline include/ignition/gui/qml/GzColor.qml:24]: Let's name this to something else like `gzColorRoot` for safety

File: include/ignition/gui/qml/GzColor.qml [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  property string colorName
+
+  property alias r: r_spin.value
+  property alias g: g_spin.value
+  property alias b: b_spin.value
+  property alias a: a_spin.value
+
+  property bool textVisible: false
```
Comment (jennuine) [inline include/ignition/gui/qml/GzColor.qml:32]: nit: `textVisible` isn't very descriptive. Can we rename this to `showColorText`?

Comment (AzulRadio) [reply include/ignition/gui/qml/GzColor.qml]: Remove `textVisible` in 8835c72

File: include/ignition/gui/qml/GzColor.qml (hunk 3 above)
Comment (AzulRadio) [reply include/ignition/gui/qml/GzColor.qml]: Changed in 8835c72

File: include/ignition/gui/qml/GzColor.qml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+*/
+import QtQuick 2.9
+import QtQuick.Controls 2.1
+import QtQuick.Dialogs 1.0
+import QtQuick.Layouts 1.3
+
+
+Item {
+  id: gzColorRoot
+
```
Comment (jennuine) [inline include/ignition/gui/qml/GzColor.qml:25]: nit: remove extra whitespace
```suggestion

```

File: include/ignition/gui/qml/GzColor.qml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import QtQuick.Layouts 1.3
+
+
+Item {
+  id: gzColorRoot
+  
+  implicitWidth: 40
+  implicitHeight: 40
+
+  property double r: 255
```
Comment (jennuine) [inline include/ignition/gui/qml/GzColor.qml:29]: Let's go ahead and update colors to be between `0 - 1.0`, that way we don't have to divide by `255` everywhere. Also, can you add a comment above `Item {` that states rgba values are between `0-1` to help implementors?

File: include/ignition/gui/qml/GzColor.qml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  property double g: 0
+  property double b: 0
+  property double a: 1.0
+
+  signal colorSet()
+
+  Button {
+    id: colorButton
```
Comment (jennuine) [inline include/ignition/gui/qml/GzColor.qml:37]:
```suggestion
    id: gzColorButton
```

File: include/ignition/gui/qml/GzColor.qml [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    Layout.leftMargin: 5
+    ToolTip.text: "Open color dialog"
+    ToolTip.visible: hovered
+    ToolTip.delay: Qt.styleHints.mousePressAndHoldInterval
+    background: Rectangle {
+      implicitWidth: 40
+      implicitHeight: 40
+      radius: 5
+      border.color: Qt.rgba(r / 255, g / 255, b / 255, a)
```
Comment (jennuine) [inline include/ignition/gui/qml/GzColor.qml:46]: Currently, if I choose white for a color the color button isn't clearly visible. Let's have the border color be black when the material theme is light and white when the theme is dark. You should be able to use something like this: https://github.com/gazebosim/gz-gui/blob/c7bfb8acf88ce6f0e46aa5bb4d214e7184e41aea/include/ignition/gui/qml/IgnSnackBar.qml#L136

File: src/plugins/grid_3d/Grid3D.qml [hunk 9]
```diff
@@ -286,98 +286,18 @@ GridLayout {
...
-      }
-    }
-  }
-
-  // Bottom spacer
-  Item {
-    Layout.columnSpan: 4
-    Layout.fillHeight: true
+  GzColor { 
+    id: gzcolor
```
Comment (jennuine) [inline src/plugins/grid_3d/Grid3D.qml:295]:
```suggestion
    id: gzColorGrid
```

**Review by jennuine — COMMENTED** [review]
Looks good so far! A couple of things,

Anytime the color dialog is open, this warning appears in the terminal:
```bash
Gtk-Message: 11:24:17.963: GtkDialog mapped without a transient parent. This is discouraged.
```
This happened before anytime the native dialog is used (but doesn't happen with Fortress' material color dialog). It would be nice to resolve it so it doesn't spam the console if a user changes colors often.

Also, when I select a custom color, the color appears twice next to the custom color section:
Peek 2022-06-16 11-22

**Comment by AzulRadio** [comment]
> Looks good so far! A couple of things,
> 
> Anytime the color dialog is open, this warning appears in the terminal:
> 
> ```shell
> Gtk-Message: 11:24:17.963: GtkDialog mapped without a transient parent. This is discouraged.
> ```
> 
> This happened before anytime the native dialog is used (but doesn't happen with Fortress' material color dialog). It would be nice to resolve it so it doesn't spam the console if a user changes colors often.
> 
> Also, when I select a custom color, the color appears twice next to the custom color section: Peek 2022-06-16 11-22 [ ](https://user-images.githubusercontent.com/8602001/174113323-0175cfd4-341e-4a34-97a4-6c3644b5696d.gif) Peek 2022-06-16 11-22 [ ](https://user-images.githubusercontent.com/8602001/174113323-0175cfd4-341e-4a34-97a4-6c3644b5696d.gif)

**Review by jennuine — COMMENTED** [review]
Thanks for iterating. A few more minor comments.

> I don't have this warning on Ogre but I think I might know how to fix this. An attempt to fix is included in https://github.com/gazebosim/gz-gui/commit/86d1b1a373fa97d9544c601d6b665c1a1dc136ae.

I didn't see the attempt to fix the warning. Was it included in that commit? The warning still appears.

> > Also, when I select a custom color, the color appears twice next to the custom color section
>
> This seems to be a bug of Qt. Even if I use the example code from their official page, this bug still persists. https://doc.qt.io/qt-5/qml-qtquick-dialogs-colordialog.html

I see, it does the same w/o this PR too. Strange, this wasn't an issue before. Anyways, thanks for looking into this

File: include/ignition/gui/qml/GzColor.qml [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  implicitHeight: 40
+
+  property double r: 1.0
+  property double g: 0.0
+  property double b: 0.0
+  property double a: 1.0
+
+  property bool useWhiteTheme: true
+
+  signal colorSet()
```
Comment (jennuine) [inline include/ignition/gui/qml/GzColor.qml:37]: Let's be more explicit and rename this signal to:
```suggestion
  signal gzColorSet()
```

**Review by jennuine — APPROVED** [review]
LGTM, nice work!

---

> Anytime the color dialog is open, this warning appears in the terminal:
> 
> ```shell
> Gtk-Message: 11:24:17.963: GtkDialog mapped without a transient parent. This is discouraged.
> ```

This message is coming from the OS's native dialog, not from this PR. The fix would be to use Qt's color dialog but there is no way to easily use that through QML currently.

**Pull request merged**